README.md
# Helix Purge

> Purge two CDNs at once

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-purge.svg)](https://codecov.io/gh/adobe/helix-purge)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-purge.svg)](https://circleci.com/gh/adobe/helix-purge)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-purge.svg)](https://github.com/adobe/helix-purge/blob/main/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-purge.svg)](https://github.com/adobe/helix-purge/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-purge.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-purge)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Installation

## Usage

```bash
curl -X POST https://adobeioruntime.net/api/v1/web/helix/helix-services/purge@v1?host=…&xfh=…&path=…
```

- `host`: the inner cdn hostname
- `xfh`: the outer cdn hostnames, comma separated
- `path`: the path to purge

For more, see the [API documentation](docs/API.md).

## Development

### Deploying Helix Purge

Deploying Helix Purge requires the `wsk` command line client, authenticated to a namespace of your choice. For Project Helix, we use the `helix` namespace.

All commits to main that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-services/purge@ci<num>` and tagged with the CI build number.

## Operations

To temporarily stop purges, edit [`OK.md`](OK.md), so that it says "not ok".

CHANGELOG.md
## [1.9.37](https://github.com/adobe/helix-purge/compare/v1.9.36...v1.9.37) (2022-02-21)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.7 ([#353](https://github.com/adobe/helix-purge/issues/353)) ([8e10718](https://github.com/adobe/helix-purge/commit/8e107181eea09e5ad9e78738f486f958352613f4))

## [1.9.36](https://github.com/adobe/helix-purge/compare/v1.9.35...v1.9.36) (2022-02-21)


### Bug Fixes

* **deps:** update adobe fixes ([#352](https://github.com/adobe/helix-purge/issues/352)) ([b961bcd](https://github.com/adobe/helix-purge/commit/b961bcd35a36f5667ba4acb2b6b829da26d913a6))

## [1.9.35](https://github.com/adobe/helix-purge/compare/v1.9.34...v1.9.35) (2022-02-21)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.5 ([#342](https://github.com/adobe/helix-purge/issues/342)) ([35e0dc0](https://github.com/adobe/helix-purge/commit/35e0dc07fd40127cfbcfd54e8bd18a54918f2fa4))

## [1.9.34](https://github.com/adobe/helix-purge/compare/v1.9.33...v1.9.34) (2022-02-04)


### Bug Fixes

* **deps:** update adobe fixes ([#341](https://github.com/adobe/helix-purge/issues/341)) ([cb85a30](https://github.com/adobe/helix-purge/commit/cb85a3014b076628d9e4c7bf4ecbafa20c87aa75))

## [1.9.33](https://github.com/adobe/helix-purge/compare/v1.9.32...v1.9.33) (2022-02-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-wrap to v1.0.1 ([#340](https://github.com/adobe/helix-purge/issues/340)) ([aa2e264](https://github.com/adobe/helix-purge/commit/aa2e26431e67085e257da6c8c74f6c345ef73546))

## [1.9.32](https://github.com/adobe/helix-purge/compare/v1.9.31...v1.9.32) (2021-11-26)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.9.1 ([#314](https://github.com/adobe/helix-purge/issues/314)) ([7dac5de](https://github.com/adobe/helix-purge/commit/7dac5de3b4d42470e4d06858b22854e0ee15b1b2))

## [1.9.31](https://github.com/adobe/helix-purge/compare/v1.9.30...v1.9.31) (2021-11-05)


### Bug Fixes

* **deps:** update dependency @adobe/fastly-native-promises to v2 ([#301](https://github.com/adobe/helix-purge/issues/301)) ([901b4fe](https://github.com/adobe/helix-purge/commit/901b4fe235295313d4af2d6e475eae37cee957ad))

## [1.9.30](https://github.com/adobe/helix-purge/compare/v1.9.29...v1.9.30) (2021-10-20)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal-logger to v2 ([#296](https://github.com/adobe/helix-purge/issues/296)) ([cae7bae](https://github.com/adobe/helix-purge/commit/cae7bae85ccea067a903bee3fa11cf9b0d508d4b))

## [1.9.29](https://github.com/adobe/helix-purge/compare/v1.9.28...v1.9.29) (2021-10-20)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.9.0 ([#294](https://github.com/adobe/helix-purge/issues/294)) ([1323d12](https://github.com/adobe/helix-purge/commit/1323d12d15858e1eeb9e0488137fbdcf965665b7))

## [1.9.28](https://github.com/adobe/helix-purge/compare/v1.9.27...v1.9.28) (2021-10-18)


### Bug Fixes

* **deps:** update adobe fixes ([#293](https://github.com/adobe/helix-purge/issues/293)) ([f430111](https://github.com/adobe/helix-purge/commit/f4301111c9ff0a5665b72a8aacc1dfd05a94ba4e))

## [1.9.27](https://github.com/adobe/helix-purge/compare/v1.9.26...v1.9.27) (2021-10-18)


### Bug Fixes

* **deps:** update dependency @adobe/helix-fetch to v3 ([#292](https://github.com/adobe/helix-purge/issues/292)) ([2dbc211](https://github.com/adobe/helix-purge/commit/2dbc211a7f1c48f24768781fe8d657f411dec782))

## [1.9.26](https://github.com/adobe/helix-purge/compare/v1.9.25...v1.9.26) (2021-09-27)


### Bug Fixes

* **deps:** use secrets-manager ([afd1f8e](https://github.com/adobe/helix-purge/commit/afd1f8e20e6fae5d54eb368ef1d5a05d2fe17579))

## [1.9.25](https://github.com/adobe/helix-purge/compare/v1.9.24...v1.9.25) (2021-09-22)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.3 ([#282](https://github.com/adobe/helix-purge/issues/282)) ([6cef82d](https://github.com/adobe/helix-purge/commit/6cef82de3e779b33583db45b704a011cac1cbbe9))

## [1.9.24](https://github.com/adobe/helix-purge/compare/v1.9.23...v1.9.24) (2021-09-17)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.12 ([#279](https://github.com/adobe/helix-purge/issues/279)) ([fd34932](https://github.com/adobe/helix-purge/commit/fd34932454ea1cfff6278a5d17414398a9862a6b))

## [1.9.23](https://github.com/adobe/helix-purge/compare/v1.9.22...v1.9.23) (2021-09-12)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.10 ([#276](https://github.com/adobe/helix-purge/issues/276)) ([2421b41](https://github.com/adobe/helix-purge/commit/2421b4192bcb2069c86fec64bc6e6a370c999bab))

## [1.9.22](https://github.com/adobe/helix-purge/compare/v1.9.21...v1.9.22) (2021-09-10)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.9 ([#274](https://github.com/adobe/helix-purge/issues/274)) ([939a3fa](https://github.com/adobe/helix-purge/commit/939a3fa095be9f4662ba57f7e6289b6bd16cfc13))

## [1.9.21](https://github.com/adobe/helix-purge/compare/v1.9.20...v1.9.21) (2021-09-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal-logger to v1.0.8 ([#271](https://github.com/adobe/helix-purge/issues/271)) ([c43d595](https://github.com/adobe/helix-purge/commit/c43d595985d7d5de4e4247c7dad0bbe1507f071a))

## [1.9.20](https://github.com/adobe/helix-purge/compare/v1.9.19...v1.9.20) (2021-09-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.8 ([#270](https://github.com/adobe/helix-purge/issues/270)) ([5c8e05a](https://github.com/adobe/helix-purge/commit/5c8e05ae16be08c7c212fb3aaf4b98f7ebbfd96e))

## [1.9.19](https://github.com/adobe/helix-purge/compare/v1.9.18...v1.9.19) (2021-08-30)


### Bug Fixes


test/utils.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const querystring = require('querystring');

function retrofit(fn) {
  const resolver = {
    createURL({ package, name, version }) {
      return new URL(`https://adobeioruntime.net/api/v1/web/helix/${package}/${name}@${version}`);
    },
  };
  return async (params = {}, env = {}) => {
    const resp = await fn({
      url: `https://helix-purge.com/purge?${querystring.encode(params)}`,
      headers: new Map(Object.entries(params.__ow_headers || {})),
    }, {
      resolver,
      env,
      // eslint-disable-next-line no-underscore-dangle
      log: params.__ow_logger,
    });
    return {
      statusCode: resp.status,
      body: await resp.json(),
      headers: resp.headers.plain(),
    };
  };
}

module.exports = { retrofit };


test/stop.test.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

process.env.HELIX_FETCH_FORCE_HTTP1 = 'true';
const nock = require('nock');
const assert = require('assert');
const { logging } = require('@adobe/helix-testutils');
const commence = require('../src/stop.js');

describe('Purge-Stop Tests', () => {
  it('Stop for unexpected results', async () => {
    const scope = nock('https://main--helix-purge--adobe.hlx.page')
      .get('/OK.plain.html')
      .reply(200, 'OK')
      .get('/ok.html')
      .reply(404, 'OK');
    assert.ok(!(await commence(logging.createTestLogger())));
    scope.done();
  });

  it('Stop for different results', async () => {
    const scope = nock('https://main--helix-purge--adobe.hlx.page')
      .get('/OK.plain.html')
      .reply(200, 'Not OK')
      .get('/ok.html')
      .reply(200, 'OK');
    assert.ok(!(await commence(logging.createTestLogger())));
    scope.done();
  });

  it('Commence for expected results', async () => {
    const scope = nock('https://main--helix-purge--adobe.hlx.page')
      .get('/OK.plain.html')
      .reply(200, 'OK')
      .get('/ok.html')
      .reply(200, 'OK');
    assert.ok(await commence(logging.createTestLogger()));
    scope.done();
  });
});


test/index.test.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

process.env.HELIX_FETCH_FORCE_HTTP1 = 'true';
const nock = require('nock');
const assert = require('assert');
const { logging } = require('@adobe/helix-testutils');
const { computeSurrogateKey } = require('@adobe/helix-shared-utils');
const { main } = require('../src/index.js');
const { retrofit } = require('./utils.js');

const index = retrofit(main);

/* eslint-disable no-underscore-dangle, camelcase */
const __ow_logger = logging.createTestLogger();

describe('Index Tests', () => {
  afterEach(() => {
    nock.cleanAll();
  });

  it('index function is present', async () => {
    const result = await index({
      __ow_logger,
    });
    assert.deepStrictEqual(result, {
      body: [],
      statusCode: 204,
      headers: {
        'content-type': 'application/json',
      },
    });
  }).timeout(10000);

  it('index function purges outer cdn', async () => {
    const scope = nock(/./)
      .intercept(/.*/, 'PURGE')
      .reply(200)
      .persist();

    const result = await index({
      __ow_logger,
      xfh: 'blog.adobe.com',
      path: '/index.html',
    });

    scope.done();
    assert.strictEqual(result.statusCode, 200);
    assert.deepStrictEqual(result.body, [
      { status: 'ok', url: 'https://blog.adobe.com/index.html' },
      { status: 'ok', url: 'https://blog.adobe.com/' },
      { status: 'ok', url: 'https://blog.adobe.com/index' },
    ]);
  }).timeout(5000);

  it('index function also purges outer cdn without html extension', async () => {
    const scope = nock(/./)
      .intercept(/\/en\/topics\/news.*/, 'PURGE')
      .twice()
      .reply(200);

    const result = await index({
      __ow_logger,
      xfh: 'blog.adobe.com',
      path: '/en/topics/news.html',
    });

    scope.done();
    assert.strictEqual(result.statusCode, 200);
    assert.deepStrictEqual(result.body, [
      { status: 'ok', url: 'https://blog.adobe.com/en/topics/news.html' },
      { status: 'ok', url: 'https://blog.adobe.com/en/topics/news' },
    ]);
  }).timeout(5000);

  it('index function also purges outer cdn with html extension if extension is missing', async () => {
    const scope = nock(/./)
      .intercept(/\/en\/topics\/creativity.*/, 'PURGE')
      .twice()
      .reply(200);

    const result = await index({
      __ow_logger,
      xfh: 'blog.adobe.com',
      path: '/en/topics/creativity',
    });

    scope.done();
    assert.strictEqual(result.statusCode, 200);
    assert.deepStrictEqual(result.body, [
      { status: 'ok', url: 'https://blog.adobe.com/en/topics/creativity' },
      { status: 'ok', url: 'https://blog.adobe.com/en/topics/creativity.html' },
    ]);
  }).timeout(5000);

  it('index function does not purge outer cdn without extension if non-html', async () => {
    const scope = nock(/./)
      .intercept('/feed.xml', 'PURGE')
      .once()
      .reply(200);

    const result = await index({
      __ow_logger,
      xfh: 'blog.adobe.com',
      path: '/feed.xml',
    });

    scope.done();
    assert.strictEqual(result.statusCode, 200);
    assert.strictEqual(result.body.length, 1);
  }).timeout(5000);

  it('index function purges outer cdn with partial failure', async () => {
    const outerScope = nock('https://blog.adobe.com')
      .intercept(/.*/, 'PURGE')
      .reply(200)
      .persist();
    const innerScope = nock('https://theblog--adobe.hlx.page')
      .intercept(/.*/, 'PURGE')
      .reply(504);

    const result = await index({
      __ow_logger,
      xfh: 'blog.adobe.com, theblog--adobe.hlx.page',
      path: '/index.html',
    });

    outerScope.done();
    innerScope.done();
    assert.strictEqual(result.statusCode, 207);
    assert.deepStrictEqual(result.body, [
      { status: 'ok', url: 'https://blog.adobe.com/index.html' },
      { status: 'ok', url: 'https://blog.adobe.com/' },
      { status: 'ok', url: 'https://blog.adobe.com/index' },
      { status: 'error', url: 'https://theblog--adobe.hlx.page/index.html' },
    ]);
  }).timeout(5000);

  [
    {
      path: '/',
      spec: '/ also purges /index and /index.html and /index.md',
      purgeUrls: [
        'https://master--theblog--adobe.hlx.page/index.md',
        'https://blog.adobe.com/',
        'https://blog.adobe.com/index',
        'https://blog.adobe.com/index.html',
        'https://theblog--adobe.hlx.page/',
        'https://theblog--adobe.hlx.page/index',
        'https://theblog--adobe.hlx.page/index.html',
      ],
    },
    {
      path: '/index.html',
      spec: '/index.html also purges / and /index and /index.md',
      purgeUrls: [
        'https://master--theblog--adobe.hlx.page/index.md',
        'https://blog.adobe.com/index.html',
        'https://blog.adobe.com/',
        'https://blog.adobe.com/index',
        'https://theblog--adobe.hlx.page/index.html',
        'https://theblog--adobe.hlx.page/',
        'https://theblog--adobe.hlx.page/index',
      ],
    },
    {
      path: '/index',
      spec: '/index also purges / and /index.html and /index.md',
      purgeUrls: [
        'https://master--theblog--adobe.hlx.page/index.md',
        'https://blog.adobe.com/index',
        'https://blog.adobe.com/',
        'https://blog.adobe.com/index.html',
        'https://theblog--adobe.hlx.page/index',
        'https://theblog--adobe.hlx.page/',
        'https://theblog--adobe.hlx.page/index.html',
      ],
    },
    {
      path: '/foo.html',
      spec: '/foo.html also purges /foo and /foo.md',
      purgeUrls: [
        'https://master--theblog--adobe.hlx.page/foo.md',
        'https://blog.adobe.com/foo.html',
        'https://blog.adobe.com/foo',
        'https://theblog--adobe.hlx.page/foo.html',
        'https://theblog--adobe.hlx.page/foo',
      ],
    },
    {
      path: '/foo',
      spec: '/foo also purges /foo.html and /foo.md',
      purgeUrls: [
        'https://master--theblog--adobe.hlx.page/foo.md',
        'https://blog.adobe.com/foo',
        'https://blog.adobe.com/foo.html',
        'https://theblog--adobe.hlx.page/foo',
        'https://theblog--adobe.hlx.page/foo.html',
      ],
    },
    {
      path: '/foo.json',
      spec: '/foo.json also purges foo.json (content proxy)',
      purgeUrls: [
        'https://blog.adobe.com/foo.json',
        'https://theblog--adobe.hlx.page/foo.json',
      ],
    },
    {
      path: '/foo.xml',
      spec: '/foo.xml purges nothing else',
      purgeUrls: [
        'https://blog.adobe.com/foo.xml',
        'https://theblog--adobe.hlx.page/foo.xml',
      ],
    },
  ].forEach(({ path, purgeUrls }) => {
    it(`index function purges outer cdn and inner cdn for ${path}`, async () => {
      const surrogateKey = computeSurrogateKey(`https://theblog--adobe.hlx.page${path}`);
      const fastlyResponse = {};
      fastlyResponse[surrogateKey] = '19940-1591821325-42118515';
      const scope = nock(/./)
        .intercept(/.*/, 'PURGE')
        .reply(200)
        .persist();

      const result = await index({
        __ow_logger,
        xfh: 'blog.adobe.com, theblog--adobe.hlx.page',
        path,
        host: 'theblog--adobe.hlx.page',
      }, {
        HLX_PAGES_FASTLY_SVC_ID: 'test-service',
        HLX_PAGES_FASTLY_TOKEN: 'dummy',
      });

      scope.done();
      assert.strictEqual(result.statusCode, 200);
      assert.strictEqual(result.body.length, purgeUrls.length, 'purged the expected number of urls');
      purgeUrls.forEach((purgeUrl) => {
        assert.ok(result.body.some((r) => r.status === 'ok' && r.url === purgeUrl), `purged url ${purgeUrl}`);
      });
    }).timeout(5000);
  });

  it('index function purges code on www.hlx.live', async () => {
    const scope = nock(/./)
      .intercept(/.*/, 'PURGE')
      .reply(200)
      .persist();

    const result = await index({
      __ow_logger,
      xfh: 'www.hlx.live, www.hlx.page',
      path: '/foo.js',
      host: 'www.hlx.page',
    });

    scope.done();
    assert.strictEqual(result.statusCode, 200);
    assert.strictEqual(result.body.length, 2, 'purged the expected number of urls');
  }).timeout(5000);

  it('index function sanitizes x-forwarded-host before purging outer cdn', async () => {
    const scope = nock(/./)
      .intercept(/.*/, 'PURGE')
      .reply(200)
      .persist();

    const result = await index({
      __ow_logger,
      xfh: 'blog.adobe.com, theblog--adobe.hlx.live, theblog--adobe.hlx.live, , theblog--adobe.hlx.page',
      path: '/index.html',
      host: 'theblog--adobe.hlx.page',
    }, {
      HLX_PAGES_FASTLY_SVC_ID: 'test-service',
      HLX_PAGES_FASTLY_TOKEN: 'dummy',
    });

    scope.done();
    assert.strictEqual(result.statusCode, 200);
    assert.strictEqual(result.body.length, 10, '11 urls purged');
  }).timeout(5000);

  it('index function gracefully handles failed markdown purge', async () => {
    const scope = nock(/./)
      .intercept('/foo.md', 'PURGE')
      .reply(500);

    const result = await index({
      __ow_logger,
      path: '/foo.html',
      host: 'theblog--adobe.hlx.page',
    }, {
      HLX_PAGES_FASTLY_SVC_ID: 'test-service',
      HLX_PAGES_FASTLY_TOKEN: 'dummy',
    });

    scope.done();
    assert.strictEqual(result.statusCode, 207);
    assert.ok(result.body.some((r) => r.status === 'error' && r.url.endsWith('.md')), 'purging markdown failed');
  }).timeout(5000);

  it('index function fails gracefully on invalid host', async () => {
    const result = await index({
      __ow_logger,
      path: '/foo.html',
      host: 'theblog.hlx.page',
    }, {
      HLX_PAGES_FASTLY_SVC_ID: 'test-service',
      HLX_PAGES_FASTLY_TOKEN: 'dummy',
    });

    assert.strictEqual(result.body.length, 1, '1 url purged');
    assert.ok(result.body.some((r) => r.status === 'error'), 'purging invalid host failed');
  }).timeout(5000);
});


src/stop.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const { equal } = require('assert');
const fetchAPI = require('@adobe/helix-fetch');

/* istanbul ignore next */
const { fetch } = process.env.HELIX_FETCH_FORCE_HTTP1
  ? fetchAPI.context({
    alpnProtocols: [fetchAPI.ALPN_HTTP1_1],
  })
  : fetchAPI;

async function commence(log) {
  // fetch options
  const options = {
    cache: 'no-store', // disable client-side cache
    headers: {
      'cache-control': 'no-cache', // prevent runtime from caching the related invocation
    },
  };
  try {
    const [mdres, htmlres] = await Promise.all([
      fetch('https://main--helix-purge--adobe.hlx.page/OK.plain.html', options),
      fetch('https://main--helix-purge--adobe.hlx.page/ok.html', options),
    ]);

    const mdtext = await mdres.text();
    const htmltext = await htmlres.text();

    equal(
      mdtext,
      htmltext,
      'Pipeline and static are in sync',
    );
    equal(mdres.status, 200, 'Pipeline works');
    equal(htmlres.status, 200, 'Static works');

    return true;
  } catch (err) {
    log.error('unable to verify pre-flight response, helix-pages may be down', err);
    return false;
  }
}

module.exports = commence;


src/index.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const wrap = require('@adobe/helix-shared-wrap');
const { logger } = require('@adobe/helix-universal-logger');
const { wrap: status } = require('@adobe/helix-status');
const { Response } = require('@adobe/helix-fetch');
const fetchAPI = require('@adobe/helix-fetch');

/* istanbul ignore next */
const { fetch } = process.env.HELIX_FETCH_FORCE_HTTP1
  ? fetchAPI.context({
    alpnProtocols: [fetchAPI.ALPN_HTTP1_1],
  })
  : fetchAPI;

function getMdInfo(host, path, log) {
  let mdPath;
  const file = path.split('/').pop() || 'index'; // use 'index' if no filename
  if (file.endsWith('.html')) {
    mdPath = path.replace(/\.html$/, '.md');
  } else if (!file.includes('.')) {
    mdPath = `${path.endsWith(file) ? path : `${path}${file}`}.md`;
  }
  if (!mdPath) {
    log.debug('Not an html document, so no markdown purging required');
    return {};
  }
  const [projectInfo] = host.split('.');
  const ghDetails = projectInfo.split('.')[0].split('--');
  const owner = ghDetails.pop();
  const repo = ghDetails.pop();
  const branch = ghDetails[0] || 'master';
  return {
    host: `${branch}--${repo}--${owner}.hlx.page`,
    path: mdPath,
  };
}

async function purge(host, path, log, exact) {
  const url = `https://${host}${path}`;
  log.info('Purging', url);
  const results = [];
  try {
    const res = await fetch(url, {
      method: 'PURGE',
    });
    const msg = await res.text();
    log.debug(msg);
    if (!res.ok) {
      throw new Error(msg);
    }
    results.push({ status: 'ok', url });
  } catch (e) {
    log.error('Unable to purge outer CDN', e);
    return { status: 'error', url };
  }
  if (!exact) {
    const file = path.split('/').pop();
    if (!file) {
      // directory, also purge index(.html)
      results.push(await purge(host, `${path}index`, log, true));
      results.push(await purge(host, `${path}index.html`, log, true));
    } else {
      if (file === 'index' || file === 'index.html') {
        // index(.html), also purge directory
        results.push(await purge(host, path.substring(0, path.lastIndexOf('/') + 1), log, true));
      }
      if (file.endsWith('.html')) {
        // file with .html extension, also purge without extension
        results.push(await purge(host, path.substring(0, path.lastIndexOf('.')), log, true));
      } else if (!file.includes('.')) {
        // file without extension, also purge with .html extension
        results.push(await purge(host, `${path}.html`, log, true));
      }
    }
  }
  return results.length === 1 ? results[0] : results;
}

async function purgeInner(host, path, log) {
  const results = [];
  // check host validity
  if (host.endsWith('.page') && host.includes('--')) {
    const { host: mdHost, path: mdPath } = getMdInfo(host, path, log);
    if (mdHost && mdPath) {
      // first purge markdown
      results.push(await purge(mdHost, mdPath, log, true));
    }
    results.push(await purge(host, path, log));
  } else if (host === 'www.hlx.page') {
    // special case
    results.push(await purge(host, path, log));
  } else {
    log.warn(`invalid inner CDN host: ${host}`);
    results.push({ status: 'error', url: `https://${host}${path}` });
  }
  return results.length === 1 ? results[0] : results;
}

/**
 * This is the main function
 * @param {Request} req The Request
 * @param {Context} context The context
 * @returns {Promise<Response>} The response
 */
async function main(req, context) {
  const { searchParams } = new URL(req.url);
  const host = searchParams.get('host');
  const path = searchParams.get('path') || '';
  const xfh = searchParams.get('xfh') || '';

  const { log } = context;

  let results = [];

  if (host) {
    results.push(await purgeInner(
      host,
      path,
      log,
    ));
  } else {
    log.warn('Not purging inner CDN due to missing host parameter');
  }
  results.push(...await Promise.all(Array.from(new Set(xfh
    .split(',')
    .map((fwhost) => fwhost.trim())
    .filter((fwhost) => !!fwhost)))
    .filter((fwhost) => fwhost !== host) // skip inner CDN host
    .map((fwhost) => purge(fwhost, path, log))));
  results = results.flat(2);

  if (results.length === 0) {
    return new Response(JSON.stringify(results), {
      status: 204,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }
  if (!results.find((r) => r.status !== 'ok')) {
    return new Response(JSON.stringify(results), {
      status: 200,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  return new Response(JSON.stringify(results), {
    status: 207,
    headers: {
      'Content-Type': 'application/json',
    },
  });
}

module.exports.main = wrap(main)
  .with(status)
  .with(logger.trace)
  .with(logger);
